src/tests/ftest/util/thread_manager.py

```diff
@@ -0,0 +1,163 @@
+                data.append("    stdout:")
+                for line in result.stdout_text.splitlines():
+                    data.append("      {}".format(line))
+                data.append("    stderr:")
+                for line in result.stderr_text.splitlines():
+                    data.append("      {}".format(line))
+            else:
+                for line in str(result).splitlines():
+                    data.append("    {}".format(line))
+            return data
+
+        info = ["Thread {} results:".format(self.id), "  args: {}".format(self.args), "  result:"]
+        if isinstance(self.result, list):
+            for this_result in self.result:
+                info.extend(get_result(this_result))
+        else:
+            info.extend(get_result(self.result))
+        return "\n".join(info)
+
+
+class ThreadManager():
+    """Class to manage running any method as multiple threads."""
+
+    def __init__(self, method, timeout=None):
+        """Initialize a ThreadManager object with the the method to run as a thread.
+
+        Args:
+            method (callable): [description]
+            timeout (int, optional): [description]. Defaults to None.
+        """
+        self.log = getLogger()
+        self.method = method
+        self.timeout = timeout
+        self.job_kwargs = []
+        self.futures = {}
+
+    @property
+    def qty(self):
+        """Get the number of threads.
+
+        Returns:
+            int: number of threads
+
+        """
+        return len(self.job_kwargs)
+
+    def add(self, **kwargs):
+        """Add a thread to run by specifying the keyword arguments for the thread method."""
+        self.job_kwargs.append(kwargs)
+
+    def run(self):
+        """Asynchronously run the method as a thread for each set of method arguments.
+
+        Returns:
+            list: a list of ThreadResults for the execution of each method.
+
+        """
+        results = []
+        with ThreadPoolExecutor() as thread_executor:
+            self.log.info("Submitting %d threads ...", len(self.job_kwargs))
+            futures = {
+                thread_executor.submit(self.method, **kwargs): index
+                for index, kwargs in enumerate(self.job_kwargs)}
+            try:
+                for future in as_completed(futures, self.timeout):
+                    id = futures[future]
+                    try:
+                        results.append(
+                            ThreadResult(id, True, self.job_kwargs[id], future.result()))
+                        self.log.info("Thread %d passed: %s", id, results[-1])
+                    except Exception as error:
+                        results.append(ThreadResult(id, False, self.job_kwargs[id], str(error)))
+                        self.log.info("Thread %d failed: %s", id, results[-1])
+            except TimeoutError as error:
+                for future in futures:
+                    if not future.done():
```

Comment:
not sure how concurrent futures works, so just making sure that if TimeoutError exception is raised, there must be at least one future in futures that is not done? not asking for anything to change here, just a question

Reply:
Correct, https://docs.python.org/3.6/library/concurrent.futures.html#concurrent.futures.as_completed